storage-node/src/services/sync/cleanupService.ts

```diff
@@ -0,0 +1,154 @@
+  const qnState = await qnApi.getQueryNodeState()
+  if (!qnState) {
+    throw new Error("Can't perform cleanup because QueryNode state info is unavailable")
+  }
+
+  const qnCurrentLag = qnState.chainHead - qnState.lastCompleteBlock
+
+  if (qnCurrentLag > MAXIMUM_QN_LAGGING_THRESHOLD) {
+    throw new Error(
+      `Can't perform cleanup as QueryNode processing is lagging by more than` +
+        `max allowed lagging threshold of ${MAXIMUM_QN_LAGGING_THRESHOLD}`
+    )
+  }
+
+  const [model, storedObjectsIds] = await Promise.all([
+    getStorageObligationsFromRuntime(qnApi, buckets),
+    getDataObjectIDs(),
+  ])
+  const assignedObjectsIds = model.dataObjects.map((obj) => obj.id)
+  const removedIds = _.difference(storedObjectsIds, assignedObjectsIds)
+  const removedObjects = await getDataObjectsByIDs(qnApi, removedIds)
+
+  logger.debug(`Cleanup - pruning ${removedIds.length} obsolete objects`)
+
+  // Data objects permanently deleted from the runtime
+  const deletedDataObjects = removedIds.filter(
+    (removedId) => !removedObjects.some((removedObject) => removedObject.id === removedId)
+  )
+
+  // Data objects no-longer assigned to current storage-node
+  // operated buckets, and have been moved to other buckets
+  const movedDataObjects = removedObjects
+
+  const workingStack = new WorkingStack()
+  const processSpawner = new TaskProcessorSpawner(workingStack, asyncWorkersNumber)
+
+  const deletionTasksOfDeletedDataObjects = await Promise.all(
+    deletedDataObjects.map((dataObject) => new DeleteLocalFileTask(uploadDirectory, dataObject))
+  )
+  const deletionTasksOfMovedDataObjects = await getDeletionTasksFromMovedDataObjects(
+    workerId,
+    uploadDirectory,
+    model,
+    movedDataObjects
+  )
+
+  await workingStack.add(deletionTasksOfDeletedDataObjects)
+  await workingStack.add(deletionTasksOfMovedDataObjects)
+  await processSpawner.process()
+  logger.info('Cleanup ended.')
+}
+
+/**
+ * Creates the local file deletion tasks.
+ *
+ * @param currentWorkerId - Worker ID
+ * @param uploadDirectory - local directory for data uploading
+ * @param dataObligations - defines the current data obligations for the node
+ * @param movedDataObjects- obsolete (no longer assigned) data objects that has been moved to other buckets
+ */
+async function getDeletionTasksFromMovedDataObjects(
+  currentWorkerId: number,
+  uploadDirectory: string,
+  dataObligations: DataObligations,
+  movedDataObjects: DataObjectDetailsFragment[]
+): Promise<DeleteLocalFileTask[]> {
+  const bucketOperatorUrlById = new Map()
+  for (const entry of dataObligations.storageBuckets) {
+    // Skip all buckets of the current WorkerId (this storage provider)
+    if (entry.workerId !== currentWorkerId) {
+      bucketOperatorUrlById.set(entry.id, entry.operatorUrl)
+    }
+  }
+
+  const timeoutMs = 60 * 1000 // 1 minute since it's only a HEAD request
+  const deletionTasks: DeleteLocalFileTask[] = []
+  await Promise.all(
```

Comment:
```suggestion
  await Promise.allSettled(
```
To avoid the promise being rejected when HEAD request to one data object fails, [Promise.allSettled()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) is more appropriate.

Also I can image if the number of dataobjects could potentially be very large, doing a massive number of HTTP HEAD requests in parallel might be a bad idea. Perhaps we should do it in batches.

Reply:
Done, but I have to change `target` option in `tsconfig.json` to `es2020` as `Promise.allSettled` does not seem to be available in `es2017`